iotile_ext_cloud/iotile/cloud/apps/ota_updater.py

```diff
@@ -0,0 +1,210 @@
+           "==": operator.ne,
+           "eq": operator.ne
+          }
+
+
+def _download_ota_script(script_url):
+    """Download the script from the cloud service and store to temporary file location"""
+
+    try:
+        urllib.request.urlretrieve(script_url, "temp.trub")
+        return "temp.trub"
+    except Exception as e:
+        print("Failed to download OTA script")
+        print(e)
+        return False
+
+
+@context("OtaUpdater")
+class OtaUpdater(IOTileApp):
+    """An IOtile app that can get OTA updates from the cloud and apply them to the device.
+    The primary function is a bundled "check_and_update" that
+
+    (1) Checks the cloud for deployment requests
+    (2) Grabs the most recent one for this device (if there are multiple)
+    (3) Checks that the update is applicable (version)
+    (4) Downloads the update script
+    (5) Attempts an update
+    (6) Informs the cloud of results
+
+    Args:
+        hw (HardwareManager): A HardwareManager instance connected to a
+            matching device.
+        app_info (tuple): The app_tag and version of the device we are
+            connected to.
+        os_info (tuple): The os_tag and version of the device we are
+            connected to.
+        device_id (int): The UUID of the device that we are connected to.
+    """
+
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, hw, app_info, os_info, device_id):
+        super(OtaUpdater, self).__init__(hw, app_info, os_info, device_id)
+
+        self.app_info = app_info
+        self.os_info = os_info
+
+        self._cloud = IOTileCloud()
+        self._con = self._hw.get(8, basic=True)
+
+        self.dev_slug = self._get_uuid()
+
+    @classmethod
+    def AppName(cls):
+        return 'ota_updater'
+
+    def _get_uuid(self):
+        device_id, = self._con.rpc(0x10, 0x08, result_format="L16x")
+        return device_id_to_slug(device_id)
+
+    @docannotate
+    def check_and_update(self):
+        """
+        Checks the cloud for an available OTA update and applies it to the device.
+        Informs the cloud of the attempt afterwards
+        Gives some diagnostic info to the user as process goes
+
+        :return: None
+        """
+
+        script = self.check_cloud_for_script()
+        if not script:
+            print("no OTA pending")
+            return
+
+        print("Applying deployment ", script[0])
+
+        print("Downloading script")
+        print(script[1])
+        filename = _download_ota_script(script[1])
+
+        if not filename:
+            print("Download of script failed for some reason")
+            return
+
+        print("Applying script")
+        status = self._apply_ota_update(filename=filename)
+
+        if status is False:
+            print("OTA update application failed")
+
+        print("Informing cloud of OTA status")
+        try:
+            self._inform_cloud(script[0], self.dev_slug, status)
+        except Exception as e:
```

Comment:
`annotated` methods designed to be invoked from the `iotile` tool should generally raise exceptions to indicate errors just like normal methods.  

The reasoning is because these methods are designed to be invoked both from the `iotile` tool as well as from scripts.  If the methods don't raise exceptions they can't be safely integrated into scripts because there's no way to know if they succeeded.  So, instead, we made the `iotile` tool handle exceptions raised from methods it calls gracefully and print them as error messages.  That way, you can just raise a normal exception when there is an issue and it will be printed without a stacktrace in the `iotile` tool and raised normally in a script.

The overall goal is to design a method that can be used interactively from the `iotile` shell without compromising its ability to be used as a normal python api from a script.